File: week-6/py/member_system/repository/mysql_repository/mysql_repository.py
from mysql.connector.pooling import MySQLConnectionPool

class MySQLRepository:
    def __init__(self, cnxpool: MySQLConnectionPool, debug: bool):
        self.debug = debug
        self.cnxpool = cnxpool
        self.createTable()

    def createTable(self):
        with self.cnxpool.get_connection() as cnx:
            with cnx.cursor() as cursor:
                if not self.tableExists():
                    cursor.execute(self.createTableStatement)

    def dropTableIfExists(self):
        if self.tableExists():
            with self.cnxpool.get_connection() as cnx:
                with cnx.cursor() as cursor:
                    cursor.execute(self.dropTableStatement)

    def tableExists(self) -> bool:
        with self.cnxpool.get_connection() as cnx:
            with cnx.cursor() as cursor:
                cursor.execute('SHOW TABLES;')
                isExists = (self.tableName,) in cursor.fetchall()
                return isExists

    @property
    def tableName(self) -> str:
        return NotImplemented

    @property
    def createTableStatement(self) -> str:
        return NotImplemented

    @property
    def dropTableStatement(self) -> str:
        return 'DROP TABLE {};'.format(self.tableName)

File: week-6/go/membersys/repository/mysql_repository.go
package repository

import (
	"database/sql"
	"fmt"

	_ "github.com/go-sql-driver/mysql"
)

type MySQLRepository struct {
	db    *sql.DB
	debug bool
}

type IMySQLRepository interface {
	CreateTableStatement() string
	TableName() string
}

func (r *MySQLRepository) CreateTable(ir IMySQLRepository) {
	if !r.TableExists(ir) {
		query := ir.CreateTableStatement()
		r.db.Query(query)
	}
}

func (r *MySQLRepository) DropTable(ir IMySQLRepository) {
	if r.TableExists(ir) {
		query := fmt.Sprintf(`DROP TABLE %s`, ir.TableName())
		r.db.Query(query)
	}
}

func (r *MySQLRepository) TableExists(ir IMySQLRepository) bool {
	query := "SHOW TABLES;"
	rows, _ := r.db.Query(query)
	for rows.Next() {
		var table string
		rows.Scan(&table)
		if table == ir.TableName() {
			return true
		}
	}
	return false
}
